File: testSuite/test-tidalTracks.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use Data::Dumper;

# Run models that test that subhalo tidal track evolution by validating against the fitting function of Errani & Navarro (2021;
# https://ui.adsabs.harvard.edu/abs/2021MNRAS.505...18E).
# Andrew Benson (17-December-2021)

# Make output directory.
system("mkdir -p outputs/");

# Iterate over models to run.
my @testCases =
    (
     {
      	 label     => "nonMonotonic",
      	 gamma     => 1.00          ,
      	 fitMetric => 0.03
     },
     {
	 label     => "monotonic",
	 gamma     => 1.0000     ,
	 fitMetric => 0.0062
     },
     {
      	 label     => "monotonic",
     	 gamma     => 0.5000     ,
	 fitMetric => 0.0056
     },
     {
      	 label     => "monotonic",
      	 gamma     => 0.000      ,
	 fitMetric => 0.028
     }
    );
foreach my $testCase ( @testCases ) {

    # Run the tidal tracks model.
    system("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/tidalTracks".ucfirst($testCase->{'label'})."_gamma".sprintf("%3.1f",$testCase->{'gamma'}).".xml");
    unless ( $? == 0 ) {
	print "FAIL: tidal track model '".$testCase->{'label'}." gamma=".sprintf("%3.1f",$testCase->{'gamma'})."' failed to run\n";
	exit;
   }

    my $velocityMaximumInitial                                                                                                                                ;
    my $radiusMaximumInitial                                                                                                                                  ;
    my $offsetMaximum          = 0.0                                                                                                                          ;
    my $model                  = new PDL::IO::HDF5("outputs/tidalTracks".ucfirst($testCase->{'label'})."_gamma".sprintf("%3.1f",$testCase->{'gamma'}).".hdf5");
    my $outputs                = $model->group('Outputs')                                                                                                     ;
    my $track;
    $track->{'model'}->{'t'} = pdl [];
    $track->{'model'}->{'r'} = pdl [];
    $track->{'model'}->{'v'} = pdl [];
    foreach my $outputName ( $outputs->groups() ) {
	my $output   = $outputs->group  ($outputName );
	my $nodeData = $output ->group  ('nodeData'  );
	my @propertyNames = ( 'time', 'nodeIsIsolated', 'darkMatterProfileDMOVelocityMaximum', 'darkMatterProfileDMORadiusVelocityMaximum' );
	my $snapshot;
	$snapshot->{$_} = $nodeData->dataset($_)->get()
	    foreach ( @propertyNames );
	my $selection = which($snapshot->{'nodeIsIsolated'} == 0);
	$track->{'model'}->{'t'} = $track->{'model'}->{'t'}->append($snapshot->{'time'                                     }->($selection));
	$track->{'model'}->{'r'} = $track->{'model'}->{'r'}->append($snapshot->{'darkMatterProfileDMORadiusVelocityMaximum'}->($selection));
	$track->{'model'}->{'v'} = $track->{'model'}->{'v'}->append($snapshot->{'darkMatterProfileDMOVelocityMaximum'      }->($selection));
    }
    my $order = $track->{'model'}->{'t'}->qsorti();
    my $r0    = $track->{'model'}->{'r'}->($order)->((0))->copy();
    my $v0    = $track->{'model'}->{'v'}->($order)->((0))->copy();
    $track->{'model'}->{'r'} /= $r0;
    $track->{'model'}->{'v'} /= $v0;
    # Construct N-body tidal track.
    if ( $testCase->{'gamma'} == 1.0 ) {
	# Tidal track from Errani & Navarro (2021).
	my $la    = pdl sequence(1000)/999.0*6.0-6.0;
	my $a     = 10.0**$la;
	my $alpha = pdl 0.40;
	my $beta  = pdl 0.65;
	$track->{'nBody'}->{'r'} = $a;
	$track->{'nBody'}->{'v'} = 2.0**$alpha*$a**$beta/(1.0+$a**2)**$alpha;
    } else {
	# Tidal tracks from Penarrubia et al. (2010).
	my $mur ;
	my $etar;
	my $muv ;
	my $etav;
	if ( $testCase->{'gamma'} == 0.0 ) {
	    $mur   = pdl -1.30;
	    $etar  = pdl +0.05;
	    $muv   = pdl +0.40;
	    $etav  = pdl +0.37;
	} elsif ( $testCase->{'gamma'} == 0.5 ) {
	    $mur   = pdl -0.40;
	    $etar  = pdl +0.27;
	    $muv   = pdl +0.40;
	    $etav  = pdl +0.35;
	} elsif ( $testCase->{'gamma'} == 1.5 ) {
	    $mur   = pdl +0.00;
	    $etar  = pdl +0.48;
	    $muv   = pdl +0.40;
	    $etav  = pdl +0.24;
	} else {
	    die("unknown gamma");
	}
	my $la    = pdl sequence(1000)/999.0*6.0-6.0;
	my $a     = 10.0**$la;
	my $r     = 2.0**$mur*$a**$etar/(1.0+$a)**$mur;
	my $v     = 2.0**$muv*$a**$etav/(1.0+$a)**$muv;
	my $order = $r->qsorti();
	$track->{'nBody'}->{'r'} = $r->($order);
	$track->{'nBody'}->{'v'} = $v->($order);
    }
    my $selection = which(($track->{'model'}->{'v'} > 0.1) & ($track->{'model'}->{'r'} > 0.1));
    my $fitMetric = 0.0;
    for(my $i=0;$i<nelem($selection);++$i) {
	my $distanceSquared =
	    +($track->{'nBody'}->{'r'}->log10()-$track->{'model'}->{'r'}->($selection)->(($i))->log10())**2
	    +($track->{'nBody'}->{'v'}->log10()-$track->{'model'}->{'v'}->($selection)->(($i))->log10())**2;
	$fitMetric += $distanceSquared->minimum();
    }
 
    my $status = $fitMetric < $testCase->{'fitMetric'} ? "SUCCESS" : "FAILED";
    print $status.": subhalo tidal tracks '".$testCase->{'label'}." gamma=".sprintf("%3.1f",$testCase->{'gamma'})."'\n";

}

exit;

File: testSuite/test-tidalTracks.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run models that test that subhalo tidal track evolution by validating against the fitting function of Errani & Navarro (2021).
# Andrew Benson (ported to Python)

# Make output directory.
subprocess.run("mkdir -p outputs/", shell=True)

testCases = [
    {"label": "nonMonotonic", "gamma": 1.00, "fitMetric": 0.0301},
    {"label": "monotonic",    "gamma": 1.00, "fitMetric": 0.0062},
    {"label": "monotonic",    "gamma": 0.50, "fitMetric": 0.0056},
    {"label": "monotonic",    "gamma": 0.00, "fitMetric": 0.0280},
]

for testCase in testCases:
    label    = testCase["label"][0].upper() + testCase["label"][1:]
    gammaStr = f"{testCase['gamma']:3.1f}"
    xmlFile  = f"testSuite/parameters/tidalTracks{label}_gamma{gammaStr}.xml"
    hdf5File = f"outputs/tidalTracks{label}_gamma{gammaStr}.hdf5"

    # Run the tidal tracks model.
    status = subprocess.run(
        f"export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe {xmlFile}",
        shell=True
    )
    if status.returncode != 0:
        print(f"FAIL: tidal track model '{testCase['label']} gamma={gammaStr}' failed to run")
        sys.exit(0)

    # Collect model track data.
    tModel = []
    rModel = []
    vModel = []
    with h5py.File(hdf5File, "r") as model:
        outputs = model["Outputs"]
        for outputName in outputs.keys():
            output   = outputs[outputName]
            nodeData = output["nodeData"]
            timeArr      = nodeData["time"][:]
            isIsolated   = nodeData["nodeIsIsolated"][:]
            vmax         = nodeData["darkMatterProfileDMOVelocityMaximum"][:]
            rmax         = nodeData["darkMatterProfileDMORadiusVelocityMaximum"][:]
            selection    = np.where(isIsolated == 0)[0]
            tModel.extend(timeArr[selection])
            rModel.extend(rmax[selection])
            vModel.extend(vmax[selection])

    tModel = np.array(tModel)
    rModel = np.array(rModel)
    vModel = np.array(vModel)
    order  = np.argsort(tModel)
    r0     = rModel[order][0]
    v0     = vModel[order][0]
    rModel = rModel / r0
    vModel = vModel / v0

    # Construct N-body tidal track.
    la = np.linspace(-6.0, 0.0, 1000)
    a  = 10.0**la

    if testCase["gamma"] == 1.0:
        # Tidal track from Errani & Navarro (2021).
        alpha          = 0.40
        beta           = 0.65
        rNBody         = a
        vNBody         = 2.0**alpha * a**beta / (1.0 + a**2)**alpha
    else:
        # Tidal tracks from Penarrubia et al. (2010).
        if testCase["gamma"] == 0.0:
            mur, etar, muv, etav = -1.30, +0.05, +0.40, +0.37
        elif testCase["gamma"] == 0.5:
            mur, etar, muv, etav = -0.40, +0.27, +0.40, +0.35
        elif testCase["gamma"] == 1.5:
            mur, etar, muv, etav = +0.00, +0.48, +0.40, +0.24
        else:
            raise ValueError(f"unknown gamma: {testCase['gamma']}")
        r      = 2.0**mur * a**etar / (1.0 + a)**mur
        v      = 2.0**muv * a**etav / (1.0 + a)**muv
        sortOrd= np.argsort(r)
        rNBody = r[sortOrd]
        vNBody = v[sortOrd]

    # Compute fit metric.
    selection = np.where((vModel > 0.1) & (rModel > 0.1))[0]
    fitMetric = 0.0
    logRNBody = np.log10(rNBody)
    logVNBody = np.log10(vNBody)
    for i in selection:
        distSq    = (logRNBody - np.log10(rModel[i]))**2 + (logVNBody - np.log10(vModel[i]))**2
        fitMetric += distSq.min()

    status_str = "SUCCESS" if fitMetric < testCase["fitMetric"] else "FAILED"
    print(f"{status_str}: subhalo tidal tracks '{testCase['label']} gamma={gammaStr}'")
